### PR TITLE
Add desc of new TiFlash storage settings and guidelines for upgrading in v4.0.9

### PR DESCRIPTION
**Pull request by JaySon-Huang**
Signed-off-by: JaySon-Huang <tshent@qq.com>

### What is changed, added or deleted? (Required)

In v4.0.9, TiFlash supports new configurations for better performance when it is deployed on multi-disk node.

This PR
* Add some description about the new configurations.
* Add guidelines for how to enable those new configurations after upgrading the TiDB cluster.

Please do NOT merge this PR until v4.0.9 is ready for release~

### Which TiDB version(s) do your changes apply to? (Required)

- [x] master (the latest development version)
- [x] v4.0 (TiDB 4.0 versions)
- [ ] v3.1 (TiDB 3.1 versions)
- [ ] v3.0 (TiDB 3.0 versions)
- [ ] v2.1 (TiDB 2.1 versions)

### What is the related PR or file link(s)?

- This PR is translated from: N/A
- Other reference link(s): https://github.com/pingcap/tiup/pull/931
- Related English version PR: https://github.com/pingcap/docs/pull/4290

### Do your changes match any of the following descriptions?

- [ ] Delete files
- [ ] Change aliases
- [x] Have version specific changes 
- [ ] Might cause conflicts

### CODE REVIEW

**Comment by TomShawn** [comment]
@JaySon-Huang Please involve a technical review~

**Comment by JaySon-Huang** [comment]
> @JaySon-Huang Please involve a technical review~

@flowbehappy

File: config-templates/complex-tiflash.yaml [hunk 1]
```diff
@@ -43,7 +43,7 @@ server_configs:
     replication.enable-placement-rules: true
   tiflash:
     # Maximum memory usage for processing a single query. Zero means unlimited.
-    profiles.default.max_memory_usage: 10000000000
```
Comment (JaySon-Huang) [inline config-templates/complex-tiflash.yaml:46]: Related PR: https://github.com/pingcap/tics/issues/1223

File: config-templates/complex-tiflash.yaml [hunk 2]
```diff
@@ -104,10 +102,35 @@ tiflash_servers:
     # flash_proxy_status_port: 20292
     # metrics_port: 8234
     # deploy_dir: /tidb-deploy/tiflash-9000
+    ## With cluster version >= v4.0.9 and you want to deploy a multi-disk TiFlash node, it is recommended to
+    ## check config.storage.* for details. The data_dir will be ignored if you defined those configurations.
+    ## Setting data_dir to a ','-joined string is still supported but deprecated.
```
Comment (flowbehappy) [inline config-templates/complex-tiflash.yaml:107]:
```suggestion
    ## The path to store data. Using ',' to separate the paths if you need multiple paths.
    ## Note that starting from v4.0.9, `data_dir` is deprecated, in favor of the new `config.storage.*` configurations. 
    ## Thought `data_dir` remains supported, it is recommended to use the new configuration format.
```

File: tiflash/tiflash-configuration.md [hunk 3]
```diff
@@ -30,25 +30,67 @@ aliases: ['/docs-cn/dev/tiflash/tiflash-configuration/','/docs-cn/dev/reference/
...
-tmp_path = tiflash 临时文件存放路径
-path = tiflash 数据存储路径     # 如果有多个目录，以英文逗号分隔
-path_realtime_mode = false # 默认为 false。如果设为 true，且 path 配置了多个目录，表示在第一个目录存放最新数据，较旧的数据存放于其他目录。
-listen_host = tiflash # TCP/HTTP 等辅助服务的监听 host，建议配置成 0.0.0.0
-tcp_port = tiflash tcp 服务端口
-http_port = tiflash http 服务端口
-mark_cache_size = 5368709120 # 数据块元信息的内存 cache 大小限制，通常不需要修改
-minmax_index_cache_size = 5368709120 # 数据块 min-max 索引的内存 cache 大小限制，通常不需要修改
-```
+# tiflash TCP/HTTP 等辅助服务的监听 host，建议配置成 0.0.0.0
```
Comment (flowbehappy) [inline tiflash/tiflash-configuration.md:33]:
```suggestion
# tiflash TCP/HTTP 等辅助服务的监听 host。建议配置成 0.0.0.0，即监听本机所有 IP 地址。
```

File: tiflash/tiflash-configuration.md [hunk 4]
```diff
@@ -30,25 +30,67 @@ aliases: ['/docs-cn/dev/tiflash/tiflash-configuration/','/docs-cn/dev/reference/
...
+
+    ## storage.main.dir 存储目录列表中，每个目录的最大可用容量。
+    ## * 在未定义配置项，或者列表中全填 0时，会使用目录所在的硬盘容量
+    ## * 以 byte 为单位，目前不支持如 "10GB" 类似的设置
+    ## * capacity 列表的长度应当与 dir 长度保持一致
+    ## 例如：
+    # capacity = [ 10737418240, 10737418240 ]
+
+    [storage.latest]
+    ## 用于存储新数据的目录列表。
```
Comment (flowbehappy) [inline tiflash/tiflash-configuration.md:78]:
```suggestion
    ## 用于存储最新的数据。该目录需要实时写入从 TiKV 同步过来的更新，故需要较高的 IOPS。
```

File: tiflash/tiflash-configuration.md [hunk 5]
```diff
@@ -30,25 +30,67 @@ aliases: ['/docs-cn/dev/tiflash/tiflash-configuration/','/docs-cn/dev/reference/
...
+[storage]
+    ## 如果节点上有多块 SSD 硬盘，推荐一个数据存储目录对应一块硬盘，并把数据目录填到列表 `storage.main.dir` 中，
+    ## 以更好地利用节点性能。
+
+    ## 如果节点上有多块 IO 性能相差较大的硬盘（如一块 SSD 硬盘加上多块 HDD 硬盘），如下配置可以更好地利用节点性能。
+    ## * 把 SSD 硬盘的数据存储目录配置在 `storage.latest.dir` 中
+    ## * 把 HDD 硬盘的数据存储目录配置在 `storage.main.dir` 中
+
+    [storage.main]
+    ## 用于存储主要数据的目录列表
```
Comment (flowbehappy) [inline tiflash/tiflash-configuration.md:65]:
```suggestion
    ## 用于存储主要的数据。该目录数据占总数据的 90% 以上。
```

File: tiflash/tiflash-configuration.md [hunk 6]
```diff
@@ -30,25 +30,67 @@ aliases: ['/docs-cn/dev/tiflash/tiflash-configuration/','/docs-cn/dev/reference/
...
+
+    [storage.latest]
+    ## 用于存储新数据的目录列表。
+    ## 在未配置或者为空列表的情况下，会使用 `storage.main.dir` 的值。
+    # dir = [ ]
+    ## storage.latest.dir 存储目录列表中，每个目录的最大可用容量。
+    # capacity = [ 10737418240, 10737418240 ]
+
+    [storage.raft]
+    ## 用于存储 Raft 数据的目录列表。
```
Comment (flowbehappy) [inline tiflash/tiflash-configuration.md:85]:
```suggestion
    ## 用于存储 Raft Log 数据，数据量很小。要实时写入从 TiKV 同步过来的数据，需要较高的 IOPS。
```

File: tiflash/tiflash-configuration.md [hunk 7]
```diff
@@ -30,25 +30,67 @@ aliases: ['/docs-cn/dev/tiflash/tiflash-configuration/','/docs-cn/dev/reference/
...
+    ## 在未定义配置项或者为空列表的情况下，会以 `storage.latest.dir` 的列表目录中每一项添加 "/kvstore" 作为此项的值。
+    # dir = [ ]
```
Comment (flowbehappy) [inline tiflash/tiflash-configuration.md:87]:
```suggestion
```

I don't think we should expose this configuration to users. They are 99.9% use the same path as `storage.latest.dir`.

File: tiflash/tiflash-configuration.md [hunk 8]
```diff
@@ -89,16 +140,111 @@ minmax_index_cache_size = 5368709120 # 数据块 min-max 索引的内存 cache 
...
+
+可以通过配置文件 [`tiflash.toml`](#配置文件-tiflashtoml) 中的 `[storage]` 项控制。下方所述的 v4.0.9 版本之前的参数，在新版本中仍然支持。但 `[storage]` 中的参数在多盘部署时，新增支持对 Raft 数据以及存储引擎中的新数据分布到多盘的特性，以便更好地利用所有硬盘性能，因此新版本中推荐使用新的配置项来配置存储路径。
+
+如果节点上有多块 SSD 硬盘，推荐一个数据存储目录对应一块硬盘，并把数据目录填到列表 `storage.main.dir` 中，以更好地利用节点性能。
+
+如果节点上有多块 IO 性能相差较大的硬盘，把 SSD 硬盘的数据存储目录配置在 `storage.latest.dir` 中，把 HDD 硬盘的数据存储目录配置在 `storage.main.dir` 中，可以更好地利用节点性能。例如 TiFlash 节点上有一块 SSD 硬盘加上两块 HDD 硬盘，你可以把 `storage.latest.dir` 设为 `["/ssd_a/data/tiflash"]` 以及把 `storage.main.dir` 设为 `["/hdd_b/data/tiflash", "/hdd_c/data/tiflash"]`。
+
+> **注意：**
+>
+> [storage] 参数在 TiUP v1.2.5 版本开始支持。请确保你的 TiUP 版本不低于 v1.2.5，否则 [storage] 中定义数据目录不会被纳入 TiUP 管理。
```
Comment (flowbehappy) [inline tiflash/tiflash-configuration.md:167]:
```suggestion
> [storage] 参数在 TiUP v1.2.5 版本开始支持。如果你的 TiDB 版本为 v4.0.9 及以上，请确保你的 TiUP 版本不低于 v1.2.5，否则 [storage] 中定义的数据目录不会被纳入 TiUP 管理。
```

File: tiflash/tiflash-configuration.md [hunk 9]
```diff
@@ -89,16 +140,111 @@ minmax_index_cache_size = 5368709120 # 数据块 min-max 索引的内存 cache 
...
 除以上几项外，其余功能参数和 TiKV 的配置相同。需要注意的是：`tiflash.toml [flash.proxy]` 中的配置项会覆盖 `tiflash-learner.toml` 中的重合参数；`key` 为 `engine` 的 `label` 是保留项，不可手动配置。
 
 ### 多盘部署
 
-TiFlash 支持多盘部署，主要通过[配置文件 `tiflash.toml`](#配置文件-tiflashtoml) 中的 `path` 和 `path_realtime_mode` 这两个参数控制。
+TiFlash 支持单节点多盘部署。如果你的部署节点上有多块硬盘，可以通过以下的方式配置参数，提高节点的硬盘IO利用率。
+
+#### 全新部署
+
+如果你部署的 TiDB 集群版本 >= v4.0.9：
```
Comment (flowbehappy) [inline tiflash/tiflash-configuration.md:157]:
```suggestion

- **如果你部署的 TiDB 集群版本 >= v4.0.9：**

```

File: tiflash/tiflash-configuration.md [hunk 10]
```diff
@@ -89,16 +140,111 @@ minmax_index_cache_size = 5368709120 # 数据块 min-max 索引的内存 cache 
...
+
+如果节点上有多块 SSD 硬盘，推荐一个数据存储目录对应一块硬盘，并把数据目录填到列表 `storage.main.dir` 中，以更好地利用节点性能。
+
+如果节点上有多块 IO 性能相差较大的硬盘，把 SSD 硬盘的数据存储目录配置在 `storage.latest.dir` 中，把 HDD 硬盘的数据存储目录配置在 `storage.main.dir` 中，可以更好地利用节点性能。例如 TiFlash 节点上有一块 SSD 硬盘加上两块 HDD 硬盘，你可以把 `storage.latest.dir` 设为 `["/ssd_a/data/tiflash"]` 以及把 `storage.main.dir` 设为 `["/hdd_b/data/tiflash", "/hdd_c/data/tiflash"]`。
+
+> **注意：**
+>
+> [storage] 参数在 TiUP v1.2.5 版本开始支持。请确保你的 TiUP 版本不低于 v1.2.5，否则 [storage] 中定义数据目录不会被纳入 TiUP 管理。
+
+如果你部署的 TiDB 集群版本 < v4.0.9：
```
Comment (flowbehappy) [inline tiflash/tiflash-configuration.md:169]:
```suggestion
- **如果你部署的 TiDB 集群版本 < v4.0.9：**

```

File: tiflash/tiflash-configuration.md [hunk 11]
```diff
@@ -71,16 +113,25 @@ minmax_index_cache_size = 5368709120 # 数据块 min-max 索引的内存 cache 
...
     size = 单个日志文件的大小
     count = 最多保留日志文件个数
+
 [raft]
-    kvstore_path = kvstore 数据存储路径 # 默认为 "{path 的第一个目录}/kvstore"
     pd_addr = pd 服务地址 # 多个地址以逗号隔开
+    ## 用于存储 Raft 数据的目录，默认为 "{path 的第一个目录}/kvstore"
+    ## 从 v4.0.9 版本开始 raft.kvstore_path 参数不再推荐使用，推荐使用 "storage.raft.dir" 配置项代替，在多盘部署的情况下能更好地利用节点性能。
+    # kvstore_path = "/tidb-data/tiflash-9000/kvstore"
+
```
Comment (flowbehappy) [inline tiflash/tiflash-configuration.md:122]:
```suggestion
```
I suggest to remove it.

Comment (flowbehappy) [inline tiflash/tiflash-configuration.md:122]:
```suggestion
```
I suggest to remove it.

File: tiflash/tiflash-configuration.md [hunk 12]
```diff
@@ -89,16 +140,111 @@ minmax_index_cache_size = 5368709120 # 数据块 min-max 索引的内存 cache 
...
+通过配置文件 [`tiflash.toml`](#配置文件-tiflashtoml) 中的 `path` 和 `path_realtime_mode` 这两个参数控制。
+
+多个数据存储目录在 `path` 中以英文逗号分隔，比如 `/ssd_a/data/tiflash,/hdd_b/data/tiflash,/hdd_c/data/tiflash`。如果你的节点上有多块硬盘，推荐一个数据存储目录对应一块硬盘，并且把性能最好的硬盘放在最前面，以更好地利用节点性能。
+
+`path_realtime_mode` 参数默认值为 false，表示数据会在所有的存储目录之间进行均衡，适用于单节点多块 SSD 硬盘的部署。
+
+`path_relatime_mode` 如果设为 true，且 `path` 配置了多个目录，表示第一个目录只会存放最新数据，较旧的数据会在其他目录之间进行均衡。适用于单个节点上一块 SSD 硬盘加上多块 HDD 硬盘的部署。
 
-多个数据存储目录在 `path` 中以英文逗号分隔，比如 `/ssd_a/data/tiflash,/hdd_b/data/tiflash,/hdd_c/data/tiflash`。如果你的环境有多块磁盘，推荐一个数据存储目录对应一块磁盘，并且把性能最好的磁盘放在最前面，以发挥所有磁盘的全部性能。
+#### 升级节点版本至不低于 v4.0.9
```
Comment (flowbehappy) [inline tiflash/tiflash-configuration.md:179]:
```suggestion
#### 升级节点版本至 v4.0.9 及以上
```

File: tiflash/tiflash-configuration.md [hunk 13]
```diff
@@ -89,16 +140,111 @@ minmax_index_cache_size = 5368709120 # 数据块 min-max 索引的内存 cache 
...
+
+v4.0.9 之前的版本中，TiFlash 只支持将存储引擎中的主要数据分布在多盘上。v4.0.9 及之后的版本中，TiFlash 支持将存储引擎的主要数据、新数据，以及 Raft 数据分布在多盘上。
```
Comment (flowbehappy) [inline tiflash/tiflash-configuration.md:181]:
```suggestion
v4.0.9 之前的版本中，TiFlash 只支持将存储引擎中的主要数据分布在多盘上。v4.0.9 及之后的版本中，TiFlash 支持将存储引擎的主要数据、新数据都分布在多盘上。
```

File: tiflash/tiflash-configuration.md (hunk 9 above)
Comment (JaySon-Huang) [reply tiflash/tiflash-configuration.md:140]: I don't think it is a good idea to define a list following with multiple lines

Comment (JaySon-Huang) [reply tiflash/tiflash-configuration.md]: Use subtitle instead of list.

File: tiflash/tiflash-configuration.md [hunk 14]
```diff
@@ -30,25 +30,55 @@ aliases: ['/docs-cn/dev/tiflash/tiflash-configuration/','/docs-cn/dev/reference/
...
+
+## tiflash 数据存储路径，如果有多个目录，以英文逗号分隔。
+## 从 v4.0.9 版本开始 path 及 path_realtime_mode 参数不再推荐使用，推荐使用 [storage] 下的配置项代替，在多盘部署的情况下能更好地利用节点性能。
+# path = "/tidb-data/tiflash-9000"
+## 或
+# path = "/ssd0/tidb-data/tiflash,/ssd1/tidb-data/tiflash,/ssd2/tidb-data/tiflash"
+## 默认为 false。如果设为 true，且 path 配置了多个目录，表示在第一个目录存放最新数据，较旧的数据存放于其他目录。
+# path_realtime_mode = false
+
+## tiflash 临时文件存放路径
```
Comment (flowbehappy) [inline tiflash/tiflash-configuration.md:52]: Can we show the default path here?

**Review by flowbehappy — APPROVED** [review]
LGTM with a minor comment.

**Comment by JaySon-Huang** [comment]
Technical review is done. Please help us revise and proofread the English version.
Note that don't merge this PR until v4.0.9 is released.

File: tiflash/tiflash-configuration.md [hunk 15]
```diff
@@ -67,38 +97,65 @@ minmax_index_cache_size = 5368709120 # 数据块 min-max 索引的内存 cache 
...
+    ## 控制 raft store 持久化 WAL 的最小间隔。通过适当增大延迟以减少 IOPS 占用，默认为 4ms，设为 0ms 则关闭该优化。
+    store-batch-retry-recv-timeout = "4ms"
 ```
 
 除以上几项外，其余功能参数和 TiKV 的配置相同。需要注意的是：`tiflash.toml [flash.proxy]` 中的配置项会覆盖 `tiflash-learner.toml` 中的重合参数；`key` 为 `engine` 的 `label` 是保留项，不可手动配置。
 
 ### 多盘部署
 
-TiFlash 支持多盘部署，主要通过[配置文件 `tiflash.toml`](#配置文件-tiflashtoml) 中的 `path` 和 `path_realtime_mode` 这两个参数控制。
+TiFlash 支持单节点多盘部署。如果你的部署节点上有多块硬盘，可以通过以下的方式配置参数，提高节点的硬盘 I/O 利用率。TiUP 中参数配置格式参照 [详细 TiFlash 配置模版](https://github.com/pingcap/docs-cn/blob/master/config-templates/complex-tiflash.yaml)
```
Comment (TomShawn) [inline tiflash/tiflash-configuration.md:138]:
```suggestion
TiFlash 支持单节点多盘部署。如果你的部署节点上有多块硬盘，可以通过以下的方式配置参数，提高节点的硬盘 I/O 利用率。TiUP 中参数配置格式参照[详细 TiFlash 配置模版](https://github.com/pingcap/docs-cn/blob/master/config-templates/complex-tiflash.yaml)。
```

File: tiflash/tiflash-configuration.md [hunk 16]
```diff
@@ -67,38 +97,65 @@ minmax_index_cache_size = 5368709120 # 数据块 min-max 索引的内存 cache 
...
+
+#### 部署的 TiDB 集群版本 v4.0.9 及以上
```
Comment (TomShawn) [inline tiflash/tiflash-configuration.md:140]:
```suggestion
#### TiDB 集群版本为 v4.0.9 及以上
```

File: tiflash/tiflash-configuration.md [hunk 17]
```diff
@@ -67,38 +97,65 @@ minmax_index_cache_size = 5368709120 # 数据块 min-max 索引的内存 cache 
...
+如果节点上有多块相同规格的硬盘，推荐把硬盘目录填到列表 `storage.main.dir` 中，`storage.latest.dir` 列表留空。TiFlash 会在所有存储目录之间分摊 I/O 压力以及进行数据均衡。
+
+如果节点上有多块规格不一致的硬盘，推荐把 I/O 性能较好的硬盘目录配置在 `storage.latest.dir` 中，把 I/O 性能较一般的硬盘目录配置在 `storage.main.dir` 中。例如节点上有一块 NVME-SSD 硬盘加上两块 SATA-SSD 硬盘，你可以把 `storage.latest.dir` 设为 `["/nvme_ssd_a/data/tiflash"]` 以及把 `storage.main.dir` 设为 `["/sata_ssd_b/data/tiflash", "/sata_ssd_c/data/tiflash"]`。TiFlash 会根据两个目录列表分别进行 I/O 压力分摊及数据均衡。注意此情况下，`storage.latest.dir` 中规划的容量大小需要占总规划容量的约 10%。
+
+> **注意：**
+>
+> * [storage] 参数在 TiUP v1.2.5 版本开始支持。如果你的 TiDB 版本为 v4.0.9 及以上，请确保你的 TiUP 版本不低于 v1.2.5，否则 [storage] 中定义的数据目录不会被 TiUP 纳入管理。
+> * 在 TiFlash 节点改为使用 [storage] 配置后，如果将集群版本降级到低于 v4.0.9，可能导致 TiFlash 部分数据丢失。
+
+#### 部署的 TiDB 集群版本低于 v4.0.9
```
Comment (TomShawn) [inline tiflash/tiflash-configuration.md:153]: 把旧方法放在新方法前面，似乎更符合逻辑。
```suggestion
#### TiDB 集群版本低于 v4.0.9
```

File: tiflash/tiflash-configuration.md [hunk 18]
```diff
@@ -67,38 +97,65 @@ minmax_index_cache_size = 5368709120 # 数据块 min-max 索引的内存 cache 
...
 除以上几项外，其余功能参数和 TiKV 的配置相同。需要注意的是：`tiflash.toml [flash.proxy]` 中的配置项会覆盖 `tiflash-learner.toml` 中的重合参数；`key` 为 `engine` 的 `label` 是保留项，不可手动配置。
 
 ### 多盘部署
 
-TiFlash 支持多盘部署，主要通过[配置文件 `tiflash.toml`](#配置文件-tiflashtoml) 中的 `path` 和 `path_realtime_mode` 这两个参数控制。
+TiFlash 支持单节点多盘部署。如果你的部署节点上有多块硬盘，可以通过以下的方式配置参数，提高节点的硬盘 I/O 利用率。TiUP 中参数配置格式参照 [详细 TiFlash 配置模版](https://github.com/pingcap/docs-cn/blob/master/config-templates/complex-tiflash.yaml)
+
+#### 部署的 TiDB 集群版本 v4.0.9 及以上
+
+v4.0.9 之前的版本中，TiFlash 只支持将存储引擎中的主要数据分布在多盘上。v4.0.9 及之后的版本中，TiFlash 支持将存储引擎的主要数据、新数据都分布在多盘上。多盘部署时，推荐使用 `[storage]` 中的参数，以更好地利用节点的 I/O 性能。但下方所述 v4.0.9 版本之前的参数仍然支持。
```
Comment (TomShawn) [inline tiflash/tiflash-configuration.md:142]: "低版本的参数仍然支持"是说 v4.0.9 以后仍然支持使用旧方法部署吗？
```suggestion
TiDB v4.0.9 及之后的版本中，TiFlash 支持将存储引擎的主要数据和新数据都分布在多盘上。多盘部署时，推荐使用 `[storage]` 中的参数，以更好地利用节点的 I/O 性能。但 TiFlash 仍然支持 [TiDB 集群版本低于 v4.0.9](#tidb-集群版本低于-v409) 中的参数。
```

File: tiflash/tiflash-configuration.md [hunk 19]
```diff
@@ -67,38 +97,65 @@ minmax_index_cache_size = 5368709120 # 数据块 min-max 索引的内存 cache 
...
+
+如果节点上有多块相同规格的硬盘，推荐把硬盘目录填到列表 `storage.main.dir` 中，`storage.latest.dir` 列表留空。TiFlash 会在所有存储目录之间分摊 I/O 压力以及进行数据均衡。
```
Comment (TomShawn) [inline tiflash/tiflash-configuration.md:144]:
```suggestion
如果节点上有多块相同规格的硬盘，推荐把硬盘目录填到 `storage.main.dir` 列表中，`storage.latest.dir` 列表留空。TiFlash 会在所有存储目录之间分摊 I/O 压力以及进行数据均衡。
```

File: tiflash/tiflash-configuration.md [hunk 20]
```diff
@@ -67,38 +97,65 @@ minmax_index_cache_size = 5368709120 # 数据块 min-max 索引的内存 cache 
...
+
+如果节点上有多块规格不一致的硬盘，推荐把 I/O 性能较好的硬盘目录配置在 `storage.latest.dir` 中，把 I/O 性能较一般的硬盘目录配置在 `storage.main.dir` 中。例如节点上有一块 NVME-SSD 硬盘加上两块 SATA-SSD 硬盘，你可以把 `storage.latest.dir` 设为 `["/nvme_ssd_a/data/tiflash"]` 以及把 `storage.main.dir` 设为 `["/sata_ssd_b/data/tiflash", "/sata_ssd_c/data/tiflash"]`。TiFlash 会根据两个目录列表分别进行 I/O 压力分摊及数据均衡。注意此情况下，`storage.latest.dir` 中规划的容量大小需要占总规划容量的约 10%。
```
Comment (TomShawn) [inline tiflash/tiflash-configuration.md:146]:
```suggestion
如果节点上有多块规格不同的硬盘，推荐把 I/O 性能较好的硬盘目录配置在 `storage.latest.dir` 中，把 I/O 性能较一般的硬盘目录配置在 `storage.main.dir` 中。例如节点上有一块 NVMe-SSD 硬盘加上两块 SATA-SSD 硬盘，你可以把 `storage.latest.dir` 设为 `["/nvme_ssd_a/data/tiflash"]` 以及把 `storage.main.dir` 设为 `["/sata_ssd_b/data/tiflash", "/sata_ssd_c/data/tiflash"]`。TiFlash 会根据两个目录列表分别进行 I/O 压力分摊及数据均衡。需要注意此情况下，`storage.latest.dir` 中规划的容量大小需要占总规划容量的约 10%。
```

File: tiflash/tiflash-configuration.md [hunk 21]
```diff
@@ -67,38 +97,65 @@ minmax_index_cache_size = 5368709120 # 数据块 min-max 索引的内存 cache 
...
+
+> **注意：**
+>
+> * [storage] 参数在 TiUP v1.2.5 版本开始支持。如果你的 TiDB 版本为 v4.0.9 及以上，请确保你的 TiUP 版本不低于 v1.2.5，否则 [storage] 中定义的数据目录不会被 TiUP 纳入管理。
```
Comment (TomShawn) [inline tiflash/tiflash-configuration.md:150]:
```suggestion
> * `[storage]` 参数从 TiUP v1.2.5 版本开始支持。如果你的 TiDB 版本为 v4.0.9 及以上，请确保你的 TiUP 版本不低于 v1.2.5，否则 `[storage]` 中定义的数据目录不会被 TiUP 纳入管理。
```

File: tiflash/tiflash-configuration.md [hunk 22]
```diff
@@ -67,38 +97,65 @@ minmax_index_cache_size = 5368709120 # 数据块 min-max 索引的内存 cache 
...
+> * 在 TiFlash 节点改为使用 [storage] 配置后，如果将集群版本降级到低于 v4.0.9，可能导致 TiFlash 部分数据丢失。
```
Comment (TomShawn) [inline tiflash/tiflash-configuration.md:151]:
```suggestion
> * 在 TiFlash 节点改为使用 `[storage]` 配置后，如果将集群版本降级到低于 v4.0.9，可能导致 TiFlash 部分数据丢失。
```

File: tiflash/tiflash-configuration.md [hunk 23]
```diff
@@ -67,38 +97,65 @@ minmax_index_cache_size = 5368709120 # 数据块 min-max 索引的内存 cache 
...
+
+#### 部署的 TiDB 集群版本 v4.0.9 及以上
+
+v4.0.9 之前的版本中，TiFlash 只支持将存储引擎中的主要数据分布在多盘上。v4.0.9 及之后的版本中，TiFlash 支持将存储引擎的主要数据、新数据都分布在多盘上。多盘部署时，推荐使用 `[storage]` 中的参数，以更好地利用节点的 I/O 性能。但下方所述 v4.0.9 版本之前的参数仍然支持。
+
+如果节点上有多块相同规格的硬盘，推荐把硬盘目录填到列表 `storage.main.dir` 中，`storage.latest.dir` 列表留空。TiFlash 会在所有存储目录之间分摊 I/O 压力以及进行数据均衡。
+
+如果节点上有多块规格不一致的硬盘，推荐把 I/O 性能较好的硬盘目录配置在 `storage.latest.dir` 中，把 I/O 性能较一般的硬盘目录配置在 `storage.main.dir` 中。例如节点上有一块 NVME-SSD 硬盘加上两块 SATA-SSD 硬盘，你可以把 `storage.latest.dir` 设为 `["/nvme_ssd_a/data/tiflash"]` 以及把 `storage.main.dir` 设为 `["/sata_ssd_b/data/tiflash", "/sata_ssd_c/data/tiflash"]`。TiFlash 会根据两个目录列表分别进行 I/O 压力分摊及数据均衡。注意此情况下，`storage.latest.dir` 中规划的容量大小需要占总规划容量的约 10%。
+
+> **注意：**
```
Comment (TomShawn) [inline tiflash/tiflash-configuration.md:148]: 涉及丢数据的内容，需要提高级别。
```suggestion
> **警告：**
```

File: tiflash/tiflash-configuration.md [hunk 24]
```diff
@@ -67,38 +97,65 @@ minmax_index_cache_size = 5368709120 # 数据块 min-max 索引的内存 cache 
...
+>
+> * [storage] 参数在 TiUP v1.2.5 版本开始支持。如果你的 TiDB 版本为 v4.0.9 及以上，请确保你的 TiUP 版本不低于 v1.2.5，否则 [storage] 中定义的数据目录不会被 TiUP 纳入管理。
+> * 在 TiFlash 节点改为使用 [storage] 配置后，如果将集群版本降级到低于 v4.0.9，可能导致 TiFlash 部分数据丢失。
+
+#### 部署的 TiDB 集群版本低于 v4.0.9
+
+通过 `path` (TiUP 中为 `data_dir`) 和 `path_realtime_mode` 这两个参数配置多盘部署。
```
Comment (TomShawn) [inline tiflash/tiflash-configuration.md:155]:
```suggestion
TiDB v4.0.9 之前的版本中，TiFlash 只支持将存储引擎中的主要数据分布在多盘上。通过 `path`（TiUP 中为 `data_dir`）和 `path_realtime_mode` 这两个参数配置多盘部署。
```

File: tiflash/tiflash-configuration.md [hunk 25]
```diff
@@ -67,38 +97,65 @@ minmax_index_cache_size = 5368709120 # 数据块 min-max 索引的内存 cache 
...
+
+多个数据存储目录在 `path` 中以英文逗号分隔，比如 `/nvme_ssd_a/data/tiflash,/sata_ssd_b/data/tiflash,/sata_ssd_c/data/tiflash`。如果你的节点上有多块硬盘，推荐把性能最好的硬盘目录放在最前面，以更好地利用节点性能。
 
-多个数据存储目录在 `path` 中以英文逗号分隔，比如 `/ssd_a/data/tiflash,/hdd_b/data/tiflash,/hdd_c/data/tiflash`。如果你的环境有多块磁盘，推荐一个数据存储目录对应一块磁盘，并且把性能最好的磁盘放在最前面，以发挥所有磁盘的全部性能。
+如果节点上有多块相同规格的硬盘，可以把 `path_realtime_mode` 参数留空（或者把该值明确地设为 `false`）。这表示数据会在所有的存储目录之间进行均衡。但由于最新的数据仍然只会被写入到第一个目录，因此其对应的硬盘会较其他硬盘繁忙。
```
Comment (TomShawn) [inline tiflash/tiflash-configuration.md:159]:
```suggestion
如果节点上有多块相同规格的硬盘，可以把 `path_realtime_mode` 参数留空（或者把该值明确地设为 `false`）。这表示数据会在所有的存储目录之间进行均衡。但由于最新的数据仍然只会被写入到第一个目录，因此该目录所在的硬盘会较其他硬盘繁忙。
```

File: tiflash/tiflash-configuration.md [hunk 26]
```diff
@@ -30,28 +30,58 @@ aliases: ['/docs-cn/dev/tiflash/tiflash-configuration/','/docs-cn/dev/reference/
...
+## 存储路径相关配置， 从 v4.0.9 开始生效
+[storage]
+    [storage.main]
+    ## 用于存储主要的数据，该目录列表中的数据占总数据的 90% 以上。
+    dir = [ "/tidb-data/tiflash-9000" ] 
+    ## 或
+    # dir = [ "/ssd0/tidb-data/tiflash", "/ssd1/tidb-data/tiflash" ]
+
+    ## storage.main.dir 存储目录列表中，每个目录的最大可用容量。
+    ## * 在未定义配置项，或者列表中全填 0 时，会使用目录所在的硬盘容量
```
[the remark this answers is not in the document]
Comment (JaySon-Huang) [reply tiflash/tiflash-configuration.md:64]: `*` 的三行是描述 storage.main.capacity 值的三个约束

File: tiflash/tiflash-configuration.md (hunk 18 above)
Comment (JaySon-Huang) [reply tiflash/tiflash-configuration.md:142]: 是的

File: tiflash/tiflash-configuration.md (hunk 26 above)
[the remark this answers is not in the document]
Comment (TomShawn) [reply tiflash/tiflash-configuration.md:64]: 所以星号表示列表吗？

File: tiflash/tiflash-configuration.md [hunk 27]
```diff
@@ -30,28 +30,58 @@ aliases: ['/docs-cn/dev/tiflash/tiflash-configuration/','/docs-cn/dev/reference/
...
-path_realtime_mode = false # 默认为 false。如果设为 true，且 path 配置了多个目录，表示在第一个目录存放最新数据，较旧的数据存放于其他目录。
-listen_host = tiflash # TCP/HTTP 等辅助服务的监听 host，建议配置成 0.0.0.0
-tcp_port = tiflash tcp 服务端口
-http_port = tiflash http 服务端口
-mark_cache_size = 5368709120 # 数据块元信息的内存 cache 大小限制，通常不需要修改
-minmax_index_cache_size = 5368709120 # 数据块 min-max 索引的内存 cache 大小限制，通常不需要修改
-```
+## TiFlash TCP/HTTP 等辅助服务的监听 host。建议配置成 0.0.0.0，即监听本机所有 IP 地址。
+listen_host = "0.0.0.0"
+## TiFlash TCP 服务的端口
```
Comment (TomShawn) [inline tiflash/tiflash-configuration.md:35]: 配置项和其他配置项的注释之间可以空一行吗

Comment (JaySon-Huang) [reply tiflash/tiflash-configuration.md:35]: 目前一些简单的配置项之间没有加空行。而是按照功能上，在相近的一些参数之间没有加空行。
担心如果全都加上空行，会不会视觉上觉得整个文件很冗长……

**Comment by TomShawn** [comment]
LGTM

File: tiflash/tiflash-configuration.md [hunk 28]
```diff
@@ -30,32 +30,61 @@ aliases: ['/docs-cn/dev/tiflash/tiflash-configuration/','/docs-cn/dev/reference/
...
+tcp_port = 9000
+## TiFlash HTTP 服务的端口
+http_port = 8123
+## 数据块元信息的内存 cache 大小限制，通常不需要修改
+mark_cache_size = 5368709120
+## 数据块 min-max 索引的内存 cache 大小限制，通常不需要修改
+minmax_index_cache_size = 5368709120
```
Comment (TomShawn) [inline tiflash/tiflash-configuration.md:42]:
```suggestion
minmax_index_cache_size = 5368709120
## DeltaIndex 内存 cache 大小限制，默认为 0，代表没有限制
delta_index_cache_size = 0
```

**Comment by TomShawn** [comment]
@JaySon-Huang Could you please resolve the conflict before this PR is merged? thanks!

**Review by TomShawn — APPROVED** [review]

**Pull request merged**